lightning/src/ln/channel.rs

```diff
@@ -2148,6 +2149,13 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 									match e {
 										ChannelError::Ignore(ref msg) => {
 											log_info!(logger, "Failed to send HTLC with payment_hash {} due to {}", log_bytes!(payment_hash.0), msg);
+											// If we fail to send here, then this HTLC should
+											// be failed backwards. Failing to send here
+											// indicates that this HTLC may keep being put back
+											// into the holding cell without ever being
+											// successfully forwarded/failed/fulfilled, causing
+											// our counterparty to eventually close on us.
+											htlcs_to_fail.push((source.clone(), *payment_hash));
```

Comment:
After the outter match we do this:
```
					if err.is_some() {
						self.holding_cell_htlc_updates.push(htlc_update);
						if let Some(ChannelError::Ignore(_)) = err {
							// If we failed to add the HTLC, but got an Ignore error, we should
							// still send the new commitment_signed, so reset the err to None.
							err = None;
						}
```

Which, IIUC, re-pushes the send onto the holding cell, even though we're about to fail it. ISTM this will means we'll try to forward that HTLC again later?

Reply:
Yikes, should've caught that. Do you think it's worth (maybe in a follow-up PR) adding to [this area](https://github.com/rust-bitcoin/rust-lightning/blob/master/lightning/src/ln/functional_test_utils.rs#L94) to make sure there's no holding cell HTLCs when a test finishes?

Reply:
Hmm, yea, that would be cool! Probably would hit a few tests that fail just cause its a new requirement but probably worth doing.